File: FirstOccuranceInAString/FirstOccuranceInAString.py
def firstOccuranceInAString(h, n):
    hL = len(h)
    nL = len(n)

    # for i in range(hL - nL +1):
    #     match = True
    #     for j in range(nL):
    #         print(f"i+j = {i+j}, j = {j}")
    #         if h[i+j] != n[j]:
    #             match = False
    #             break
    #     if match:
    #         print("i = ", i)
    #         return i
    # return -1

    res = []
    for id in range(hL-nL+1):
        sliced = h[id:nL+id]
        if sliced == n:
            res.append(id)
            return id 

    return -1  
            
        # sliced = h[0:3] = "sad"
        # if sliced == n
        #   result = h[0]
        # sliced = h[1:4]
        #   if sliced == n

h="sadbutsad"
n="sad"
print(f"{h}, {n}, {firstOccuranceInAString(h, n)}")

h="sadbutsalt"
n="sad"
print(f"{h}, {n}, {firstOccuranceInAString(h, n)}")

h="butsadbut"
n="sad"
print(f"{h}, {n}, {firstOccuranceInAString(h, n)}")

h="nothingbuthappy"
n="sad"
print(f"{h}, {n}, {firstOccuranceInAString(h, n)}")

h="saltychill"
n="sad"
print(f"{h}, {n}, {firstOccuranceInAString(h, n)}")

File: FirstOccuranceInAString/FirstOccuranceInAString.go
package main

import "fmt"

// // Using slicing, T.C. O(n)
// func strStr(haystack string, needle string) int {
// 	nL := len(needle)
// 	hL := len(haystack)
// 	for i := 0; i < (hL - nL + 1); i++ {
// 		sliced := haystack[i : nL+i]
// 		if sliced == needle {
// 			return i
// 		}
// 	}
// 	return -1
// }

func strStr(haystack string, needle string) int {
	var res int
	np := 0
	for hp := 0; hp < len(haystack)-len(needle)+4; hp++ {
		if np < len(needle)-1 {
			if needle[np] != haystack[hp] {
				np = 0

			} else {
				np += 1
			}
		}
	}
	return res
}

func main() {
	var n, h string

	h = "sadbutsad"
	n = "sad"
	fmt.Println(h, n, strStr(h, n))

	h = "sadbutsalt"
	n = "sad"
	fmt.Println(h, n, strStr(h, n))

	h = "butsadbut"
	n = "sad"
	fmt.Println(h, n, strStr(h, n))

	h = "nothingbuthappy"
	n = "sad"
	fmt.Println(h, n, strStr(h, n))

	h = "saltychill"
	n = "sad"
	fmt.Println(h, n, strStr(h, n))
}
